dsgrid/query/query_submitter.py

```diff
@@ -0,0 +1,367 @@
+from dsgrid.utils.spark import read_dataframe, write_dataframe_and_auto_partition
+from dsgrid.utils.timing import timer_stats_collector, track_timing
+from dsgrid.utils.files import dump_data, load_data
+from .models import (
+    ChainedAggregationModel,
+    ProjectQueryResultModel,
+    CreateDerivedDatasetQueryModel,
+    DerivedDatasetQueryModel,
+    QueryResultBaseModel,
+)
+from .query_context import QueryContext
+
+
+logger = logging.getLogger(__name__)
+
+
+class QuerySubmitterBase:
+    """Handles query submission"""
+
+    def __init__(self, output_dir: Path):
+        self._output_dir = output_dir
+        self._cached_tables_dir().mkdir(exist_ok=True, parents=True)
+        self._derived_datasets_dir().mkdir(exist_ok=True, parents=True)
+
+    @abc.abstractmethod
+    def submit(self, *args, **kwargs):
+        """Submit a query for execution"""
+
+    @track_timing(timer_stats_collector)
+    def _persist_intermediate_result(self, context: QueryContext, df):
+        hash, text = context.model.serialize_cached_content()
+        cached_dir = self._cached_tables_dir() / hash
+        if cached_dir.exists():
+            shutil.rmtree(cached_dir)
+        cached_dir.mkdir()
+        filename = self._cached_table_filename(cached_dir)
+        df = write_dataframe_and_auto_partition(df, filename)
+        # TODO: These columns should probably be embedded inside the Parquet or have the
+        # dimension types encoded, such as 'metric__electricity.'
+        metadata = {
+            "metric_columns": context.metric_columns,
+            "required_dimension_mappings": context.required_dimension_mappings,
+        }
+        dump_data(metadata, self._metadata_filename(cached_dir), indent=2)
+        self._query_filename(cached_dir).write_text(text)
+        logger.info("Persisted intermediate table to %s", filename)
+        return df
+
+    def _try_read_cache(self, context: QueryContext):
+        hash, _ = context.model.serialize_cached_content()
+        cached_dir = self._cached_tables_dir() / hash
+        if cached_dir.exists():
+            filename = self._cached_table_filename(cached_dir)
+            logger.info("Load intermediate table from cache: %s", filename)
+            metadata = load_data(self._metadata_filename(cached_dir))
+            return (
+                read_dataframe(filename),
+                metadata["metric_columns"],
+                metadata["required_dimension_mappings"],
+            )
+        return None, None, None
+
+
+class ProjectBasedQuerySubmitter(QuerySubmitterBase):
+    def __init__(self, project, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._project = project
+
+    def _process_metric_reductions(self, context, df):
+        assert (
+            len(context.model.metric_reductions) <= 1
+        ), "More than one metric aggregation is not supported yet"
+        cols = set(context.metric_columns)  # TODO DT: same as pivot columns?
```

Comment:
Maybe to avoid confusion, use the name value_columns instead of metric_columns? These columns represent some values or quantities but the resolution and column names are dictated by the pivoted dimension.

Reply:
I'm open. If the subsector dimension is pivoted I could be OK with calling those the "metric columns." I could also go with "value_columns;" it just feels a bit generic. Maybe we should stick with "pivoted_columns," as stated in my note.

Reply:
Tied to reductions of the dimension that contains metric values.